Makes the MC avoid triggering the byond bug that causes lag.

MrsTonedOne opened the pull request:
See http://www.byond.com/forum/?post=2021963#comment18164837

Long story short, if world.cpu ever goes over 120%, byond suspends all client updates for world.fps*1.6 amount of ticks (25 ticks in the case of sybil's 16fps)

If the cpu is above 100%, we set world.fps to 10 and reset it back to normal once the cpu gets below 50%

(Higher fps when cpu goes over 120 causes the bug to get worst. I have it scaling around 100 to give us a buffer, and not going back down until it drops below 50 to avoid it jumping around too much.)

The mc will also scale how often it runs with world.cpu, doubling its normal delay between fires for every 10% above 50% if the cpu is above 75%, and doubling everything again if it's above 100%

The mc will not run any subsystems if at any point world.cpu is above 100%, this might come back to bite me, but it's better than byond delaying all client updates for 1.6 to 3 seconds.

This is a bit of a mess, but all of it can removed once the bug is fixed on byond's end.

I've also had to disable it trying to make up missed subsystem ticks on dynamic wait subsystems, as that was coming into play against the high cpu delays with air and lighting, and that didn't make any sense to do.

And it will be a bit less aggressive about making up missed ticks, making them up 25% at a time, rather than 50%

neersighted commented on code/controllers/master.dm:
```diff
@@ -167,7 +168,16 @@ var/global/datum/controller/master/Master = new()
+					oldfps = world.fps
+					//byond bug, if we go over 120 fps and world.fps is higher then 10, the bad things that happen are made worst.
+					world.fps = 10
+				else if (oldfps && world.cpu < 80)
```
`else if(`

neersighted commented on code/controllers/master.dm:
```diff
@@ -21,7 +21,8 @@ var/global/datum/controller/master/Master = new()
 	// The cost (in deciseconds) of the MC loop.
 	var/cost = 0
-
+	//the old fps when we slow it down to prevent lag
```
capitalization, punctuation,,, speling, and w h i t e space are all important

Fox-McCloud commented:
.....weird.

Why BYOND does some of the things it does is ....yeah.

MrStonedOne commented:
@Fox-McCloud 

Byond uses world.cpu to determine when to trigger this.

So we use it to avoid triggering it

````
//if the cpu is overloaded, punish SendMaps() for its share of the guilt
if(tick_usage > 120 && map_tick_usage > tick_usage - 100) {
    map_countdown = map_tick_usage;
    if(map_countdown > 16) map_countdown = 16;
    map_countdown = (unsigned short)(map_countdown * 100 / tick_lag); //slow things down even more on high frequency tickers
}
else map_countdown = 0;
````

MrsTonedOne committed:
Makes the MC avoid triggering the byond bug that causes lag.

See http://www.byond.com/forum/?post=2021963#comment18164837

Higher fps when cpu goes over 120 causes issues. I have it scaling around 100 to give us a buffer, and not going back down until it drops below 75 to avoid it jumping too much.

This is a bit of a mess, but all of it can removed once the bug is fixed on byond's end.

MrPerson commented on code/controllers/master.dm:
```diff
@@ -127,6 +128,12 @@ var/global/datum/controller/master/Master = new()
+						//byond will pause most client updates for (about) 1.6 seconds.
+						//(1.6 seconds worth of ticks)
+						//We just stop running subsystems to avoid that.
+						break;
```
>\>using a semicolon as an eol

How sad is it that the best tool we have to avoid lag is to lag things up manually? Either way :+1: looks ok

MrStonedOne replied:
Doing that is habit from php/c#/c++, I don't know why, but i always do that with continue and break but nothing else.

PKPenguin321 commented:
Thanks MSO, you lovely man
I knew you could fix it

lzimann commented:
Is the FPS lower than what we use at roundstart intended? It will start the round at 11~ fps and then return to 16 after half a minute or so

LummoxJR commented:
Fox, IIRC Volundr's library was partly intended as a stopgap until something like a tick completion percentage measurement could be achieved. In 510, world.tick_usage will be included. It's not a perfect solution to anything and may be hard to use with any accuracy, but it should be helpful for situations like the ones Volundr is handling.

MrStonedOne commented:
@lzimann yes, 10 is the base byond uses to calculate.

MrsTonedOne committed:
Make up missed SS ticks less aggressively.

Don't attempt to make up missed subsystem ticks on dynamic wait subsystem

Only make them up 25% at a time.

(Also tweaks high cpu delaying a bit)

Fox-McCloud commented:
@LummoxJR 

Yep, indeed he did; we utilize that, but it's a costly thing to do so--should be interesting to see how tick_usage handles.
